libs/cardano-ledger-core/internal/Cardano/Ledger/Internal/Definition/Era.hs

```diff
@@ -10,16 +22,162 @@ module Cardano.Ledger.Internal.Definition.Era (
   DijkstraEra,
 ) where
 
+import Cardano.Ledger.Binary (MaxVersion, MinVersion)
+import Data.Kind (Type)
+import Data.Typeable (Typeable)
+import GHC.Exts (Proxy#, proxy#)
+import GHC.TypeLits
+
+--------------------------------------------------------------------------------
+-- Era
+--------------------------------------------------------------------------------
+
+class
+  ( Typeable era
+  , KnownNat (ProtVerLow era)
+  , KnownNat (ProtVerHigh era)
+  , ProtVerLow era <= ProtVerHigh era
+  , MinVersion <= ProtVerLow era
+  , MinVersion <= ProtVerHigh era
+  , -- We need to make sure that there is never a case that MaxVersion equals to the highest
+    -- protocol version for the era, otherwise we can never upgrade to the next version:
+    CmpNat (ProtVerLow era) MaxVersion ~ 'LT
+  , CmpNat (ProtVerHigh era) MaxVersion ~ 'LT
+  ) =>
+  Era era
+  where
+  type EraName era :: Symbol
+
+  -- | Map an era to its predecessor.
+  --
+  -- For example:
+  --
+  -- > type instance PreviousEra (AllegraEra c) = ShelleyEra c
```

Comment:
```suggestion
  -- > type instance PreviousEra AllegraEra = ShelleyEra
```
Is this applicable now that we do not have a parameter for eras the way we had crypto earlier?

Reply:
Nice catch. Thank you!